tools/smithy-rs-sync/src/main.rs

```diff
@@ -0,0 +1,409 @@
+/// need to sync. If you don't see the commits you're expecting, make sure the repo is up to date.
+/// This function doesn't include the `since_commit` in the list since that commit was synced last time
+/// this tool was run.
+fn commits_to_be_applied(smithy_rs_repo: &Repository, since_commit: &Oid) -> Result<Vec<Oid>> {
+    let rev_range = format!("{}..HEAD", since_commit);
+    println!("Checking for smithy-rs commits in range {}", rev_range);
+
+    let mut rev_walk = smithy_rs_repo.revwalk().context(here!())?;
+    rev_walk.push_range(&rev_range).context(here!())?;
+
+    let mut commit_revs = Vec::new();
+    for rev in rev_walk {
+        let rev = rev.context(here!())?;
+        commit_revs.push(rev);
+    }
+
+    // Order the revs from earliest to latest
+    commit_revs.reverse();
+
+    Ok(commit_revs)
+}
+
+/// Read the file from aws-sdk-rust that tracks the last smithy-rs commit it was synced with.
+/// Returns the hash of that commit.
+fn get_last_synced_commit(repo_path: &Path) -> Result<Oid> {
+    let path = repo_path.join(COMMIT_HASH_FILENAME);
+
+    let mut file = OpenOptions::new()
+        .read(true)
+        .open(&path)
+        .with_context(|| format!("couldn't open file at '{}'", path.to_string_lossy()))?;
+    // Commit hashes are 40 chars long
+    let mut commit_hash = String::with_capacity(40);
+    file.read_to_string(&mut commit_hash)
+        .with_context(|| format!("couldn't read file at '{}'", path.to_string_lossy()))?;
+
+    // We trim here in case some really helpful IDE added a newline to the file
+    let oid = Oid::from_str(commit_hash.trim()).context(here!())?;
+
+    Ok(oid)
+}
+
+/// Write the last synced commit to the file in aws-sdk-rust that tracks the last smithy-rs commit it was synced with.
+fn set_last_synced_commit(repo: &Repository, oid: &Oid) -> Result<()> {
+    let repo_path = repo.workdir().expect("this will always exist");
+    let path = repo_path.join(COMMIT_HASH_FILENAME);
+    let mut file = OpenOptions::new().write(true).truncate(true).open(&path)?;
+
+    file.write(oid.to_string().as_bytes())
+        .with_context(|| format!("Couldn't write commit hash to '{}'", path.to_string_lossy()))?;
+
+    Ok(())
+}
+
+/// Run the necessary commands to build the SDK. On success, returns the path to the folder containing
+/// the build artifacts.
+fn build_sdk(smithy_rs_path: &Path) -> Result<PathBuf> {
+    println!("\tbuilding the SDK...");
+    let start = Instant::now();
+    let gradlew = smithy_rs_path.join("gradlew");
+
+    // The output of running this command isn't logged anywhere unless it fails
+    let clean_command_output = Command::new(&gradlew)
+        .arg("-Paws.fullsdk=true")
+        .arg(":aws:sdk:clean")
+        .current_dir(smithy_rs_path)
+        .output()
+        .with_context(|| {
+            format!(
+                "failed to execute '{} -Paws.fullsdk=true :aws:sdk:clean'",
+                gradlew.to_string_lossy()
+            )
+        })?;
```

Comment:
probably worth adding something like:

`fn run(args: &[str]) -> ...` that runs the command and handles the error wrapping